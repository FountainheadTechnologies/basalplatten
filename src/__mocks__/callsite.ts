module.exports = () => ([{
}, {
  getFileName: () => 'test-caller.js'
}]);
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
