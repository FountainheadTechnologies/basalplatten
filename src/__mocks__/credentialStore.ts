export const credentialStore = {
  setToken: jest.fn(),
  getToken: jest.fn(),
  clearToken: jest.fn()
}
