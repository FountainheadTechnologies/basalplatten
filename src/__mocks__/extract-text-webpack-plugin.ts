module.exports = class ExtractTextPlugin {
  extract() {}
};
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
