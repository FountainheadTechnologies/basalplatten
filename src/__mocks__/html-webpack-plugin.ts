module.exports = class HtmlWebpackPlugin {};
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
