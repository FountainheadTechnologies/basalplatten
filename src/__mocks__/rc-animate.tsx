import React from 'react';

const RcAnimate = props => (
  <div>{props.component}</div>
)

RcAnimate.displayName = 'RcAnimate';

export default RcAnimate;
