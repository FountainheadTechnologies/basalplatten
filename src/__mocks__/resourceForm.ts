const submit = jest.fn();

export const createToken = {
  schema: {
    required: ['username', 'password'],
    properties: {
      username: { type: 'string' },
      password: { type: 'string' }
    },
  },

  submit
}

export const createPerson = {
  schema: {
    required: ['forename', 'surname'],
    properties: {
      forename: { type: 'string' },
      surname: { type: 'string' },
      date_of_birth: {
        type: 'string',
        format: 'date-time'
      }
    }
  },

  submit
}

export const updatePerson = {
  schema: {
    required: [],
    default: {
      forename: 'Test',
      surname: 'Person',
      date_of_birth: '1985-08-02T23:00:00.000Z',
      credits: 33,
      favourite_colour: 'blue'
    },
    properties: {
      forename: { type: 'string' },
      surname: { type: 'string' },
      credits: { type: 'number' },
      enabled: { type: 'boolean' },
      date_of_birth: {
        type: 'string',
        format: 'date-time'
      },
      favourite_colour: {
        enum: ['red', 'green', 'blue']
      }
    }
  }
}
