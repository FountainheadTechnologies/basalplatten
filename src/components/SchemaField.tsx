import React from 'react';
import PropTypes from 'prop-types';
import titleCase from 'title-case';
import moment from 'moment';

import AntForm from 'antd/lib/form';
import Input from 'antd/lib/input';
import 'antd/lib/input/style';
import Switch from 'antd/lib/switch';
import 'antd/lib/switch/style';

import { WrappedFormUtils, GetFieldDecoratorOptions } from 'antd/lib/form/Form';
import { Property } from '@optics/hal-client/dist/Form';

// @hack: Prevent TS/Babel from mangling reference to constructor
const _switch = Switch;

import { FormItemProps as AntFormItemProps } from 'antd/lib/form/FormItem';

import { Context as FormContext, childContextTypes } from './Form';

export interface Props extends AntFormItemProps {
  name: string;
  showLabel?: boolean;
  defaultValue?: any;
  label?: string;
}

interface Context extends FormContext {
  form: WrappedFormUtils;
}

export const SchemaField: React.StatelessComponent<Props> = (props, context: Context) => {
  const {
    name,
    children,
    showLabel = true,
    defaultValue,
    ...itemProps
  } = props;

  const { form, schema, defaultItemProps } = context;

  const property = schema.properties[name];
  const element = children || inputElement(property);
  const label = itemProps.label || titleCase(name);

  const decorate = form.getFieldDecorator(name, fieldDecoratorOptions(props, context, element, label));

  return (
    <AntForm.Item
      {...defaultItemProps}
      {...itemProps}
      label={showLabel ? label : undefined}
      key={name}
    >
      {decorate(element)}
    </AntForm.Item>
  );
}

const inputElement = (property: Property): React.ReactNode => (
  <Input />
);

const fieldDecoratorOptions =
  (props: Props, context: Context, element: React.ReactNode, label?: string): GetFieldDecoratorOptions => {
    const { name, defaultValue, required } = props;
    const { schema } = context;

    const property = schema.properties[name];

    let initialValue = defaultValue || (schema.default && schema.default[name]);
    if (initialValue && property.format === 'date-time') {
      initialValue = moment(initialValue);
    }

    let valuePropName = 'value';
    if (element && (element as React.ReactElement<any>).type === _switch) {
      valuePropName = 'checked';
    }

    const rules = [];
    if (schema.required.indexOf(name) > -1 || required === true) {
      rules.push({
        required: true,
        message: `${label} is required`
      });
    }

    return { initialValue, valuePropName, rules };
  }

SchemaField.contextTypes = {
  ...childContextTypes,
  form: PropTypes.object
}
