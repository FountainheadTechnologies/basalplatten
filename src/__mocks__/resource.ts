import { createToken } from './resourceForm';

export const api = {
  properties: {},
  meta: {},

  config: {
    headers: {}
  },

  formNamed(rel: string, name: string) {
    if (rel === 'token' && name === 'create') {
      return createToken;
    }
  }
}
