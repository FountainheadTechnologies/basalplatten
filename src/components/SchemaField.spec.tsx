import React from 'react';
import moment from 'moment';
import { mount } from 'enzyme';
import { Switch } from 'antd';

import { SchemaField } from './SchemaField';
import { createPerson, updatePerson } from '../__mocks__/resourceForm';

const mockForm = {
  _decorate: jest.fn(element => element),
  getFieldDecorator: jest.fn(() => mockForm._decorate)
}

const createContext = {
  form: mockForm,
  schema: createPerson.schema
}

const updateContext = {
  form: mockForm,
  schema: updatePerson.schema
}

describe('default behaviour', () => {
  const schemaField = mount(<SchemaField name="forename" />, { context: createContext });

  it('renders an Ant Form Item component with an Input element', () => {
    expect(schemaField).toMatchSnapshot();
    expect(schemaField.find('Input').exists()).toBe(true);
  });

  it('adds rules on Ant Form field decorator for required fields', () => {
    expect(mockForm.getFieldDecorator).toHaveBeenCalledWith('forename', {
      initialValue: undefined,
      valuePropName: 'value',
      rules: [{
        required: true,
        message: 'Forename is required'
      }]
    })
  });
});

describe('when field is not required in schema', () => {
  const schemaField = mount(<SchemaField name="date_of_birth" />, { context: createContext });

  it('does not add Ant Form field decorator rules', () => {
    expect(mockForm.getFieldDecorator).toHaveBeenCalledWith('date_of_birth', {
      initialValue: undefined,
      valuePropName: 'value',
      rules: []
    })
  });
});

describe('when `required` prop is set', () => {
  const schemaField = mount((
    <SchemaField
      name="date_of_birth"
      required={true}
    />
  ), { context: createContext });

  it('adds Ant Form field decorator rule, even if not required in schema', () => {
    expect(mockForm.getFieldDecorator).toHaveBeenCalledWith('date_of_birth', {
      initialValue: undefined,
      valuePropName: 'value',
      rules: [{
        required: true,
        message: 'Date Of Birth is required'
      }]
    });
  });
});

describe('when default value for field is in schema', () => {
  const schemaField = mount(<SchemaField name="forename" />, { context: updateContext });

  it('sets `initialValue` in field decorator options', () => {
    expect(mockForm.getFieldDecorator).toHaveBeenCalledWith('forename', {
      initialValue: 'Test',
      valuePropName: 'value',
      rules: []
    })
  });
});

describe('when field type in schema is `date-time`', () => {
  const schemaField = mount((<SchemaField name="date_of_birth" />), { context: updateContext });

  it('sets `initialValue` to a moment instance', () => {
    expect(mockForm.getFieldDecorator).toHaveBeenCalledWith('date_of_birth', {
      initialValue: jasmine.any(moment),
      valuePropName: 'value',
      rules: []
    });
  });
});

describe('when child element is an Ant Switch', () => {
  const schemaField = mount((
    <SchemaField
      name="enabled"
    >
      <Switch />
    </SchemaField>
  ), { context: updateContext });

  it('sets `valuePropName` to `checked`', () => {
    expect(mockForm.getFieldDecorator).toHaveBeenCalledWith('enabled', {
      initialValue: undefined,
      valuePropName: 'checked',
      rules: []
    })
  });
});

describe('when `props.showLabel` is false', () => {
  const schemaField = mount((
    <SchemaField
      name="date_of_birth"
      showLabel={false}
    />
  ), { context: createContext });

  it('does not set the label on the Ant Form Item element', () => {
    expect(schemaField.find('FormItem').prop('label')).toBeUndefined();
  });
});

describe('when schema type is `number`', () => {
  const schemaField = mount(<SchemaField name="credits" />, { context: updateContext });

  it('renders an `<InputNumber/>` element', () => {
    expect(schemaField.find('InputNumber').exists()).toBe(true);
  });
});

describe('when schema type is `date-time`', () => {
  const schemaField = mount(<SchemaField name="date_of_birth" />, { context: updateContext });

  it('renders a `<DatePicker/>` element', () => {
    expect(schemaField.find('PickerWrapper').exists()).toBe(true);
  });
});

describe('when schema type is `boolean`', () => {
  const schemaField = mount(<SchemaField name="enabled" />, { context: updateContext });

  it('renders a `<Switch/>` element', () => {
    expect(schemaField.find('Switch').exists()).toBe(true);
  });
});

describe('when schema type is `enum`', () => {
  const schemaField = mount(<SchemaField name="favourite_colour" />, { context: updateContext });

  it('renders a `<Select/>` element containing possible values', () => {
    expect(schemaField.find('Select').exists()).toBe(true);
    expect(schemaField.find('Select')).toMatchSnapshot();
  });
});
