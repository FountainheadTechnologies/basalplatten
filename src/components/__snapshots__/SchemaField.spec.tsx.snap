// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default behaviour renders an Ant Form Item component with an Input element 1`] = `
<Component
  name="forename"
>
  <FormItem
    colon={true}
    hasFeedback={false}
    key="forename"
    label="Forename"
    prefixCls="ant-form"
  >
    <Row
      className="ant-form-item"
      gutter={0}
    >
      <div
        className="ant-row ant-form-item"
      >
        <Col
          className="ant-form-item-label"
          key=".$label"
        >
          <div
            className="ant-form-item-label"
          >
            <label
              className=""
              htmlFor={null}
              onClick={[Function]}
              title="Forename"
            >
              Forename
            </label>
          </div>
        </Col>
        <Col
          className="ant-form-item-control-wrapper"
          key=".$wrapper"
        >
          <div
            className="ant-form-item-control-wrapper"
          >
            <div
              className="ant-form-item-control "
            >
              <Input
                disabled={false}
                key=".0"
                prefixCls="ant-input"
                size="large"
                type="text"
              >
                <input
                  className="ant-input ant-input-lg"
                  disabled={false}
                  onKeyDown={[Function]}
                  size="large"
                  type="text"
                />
              </Input>
            </div>
          </div>
        </Col>
      </div>
    </Row>
  </FormItem>
</Component>
`;

exports[`when schema type is \`enum\` renders a \`<Select/>\` element containing possible values 1`] = `
Array [
  <Select
    choiceTransitionName="zoom"
    key=".0"
    prefixCls="ant-select"
    showSearch={false}
    size="large"
    transitionName="slide-up"
  >
    <Select
      allowClear={false}
      backfill={false}
      choiceTransitionName="zoom"
      className="ant-select-lg"
      defaultActiveFirstOption={true}
      defaultOpen={false}
      dropdownMatchSelectWidth={true}
      dropdownMenuStyle={Object {}}
      dropdownStyle={Object {}}
      labelInValue={false}
      notFoundContent="无匹配结果"
      onBlur={[Function]}
      onChange={[Function]}
      onDeselect={[Function]}
      onFocus={[Function]}
      onSearch={[Function]}
      onSelect={[Function]}
      optionFilterProp="value"
      optionLabelProp="children"
      placeholder=""
      prefixCls="ant-select"
      showArrow={true}
      showSearch={false}
      transitionName="slide-up"
    >
      <SelectTrigger
        defaultActiveFirstOption={true}
        dropdownMatchSelectWidth={true}
        dropdownMenuStyle={Object {}}
        dropdownStyle={Object {}}
        inputValue=""
        onDropdownVisibleChange={[Function]}
        onMenuDeselect={[Function]}
        onMenuSelect={[Function]}
        onPopupFocus={[Function]}
        options={Array []}
        prefixCls="ant-select"
        showSearch={false}
        transitionName="slide-up"
        value={Array []}
        visible={false}
      >
        <Trigger
          action={Array []}
          afterPopupVisibleChange={[Function]}
          blurDelay={0.15}
          builtinPlacements={
            Object {
              "bottomLeft": Object {
                "offset": Array [
                  0,
                  4,
                ],
                "overflow": Object {
                  "adjustX": 0,
                  "adjustY": 1,
                },
                "points": Array [
                  "tl",
                  "bl",
                ],
              },
              "topLeft": Object {
                "offset": Array [
                  0,
                  -4,
                ],
                "overflow": Object {
                  "adjustX": 0,
                  "adjustY": 1,
                },
                "points": Array [
                  "bl",
                  "tl",
                ],
              },
            }
          }
          defaultActiveFirstOption={true}
          defaultPopupVisible={false}
          destroyPopupOnHide={false}
          dropdownMatchSelectWidth={true}
          dropdownMenuStyle={Object {}}
          dropdownStyle={Object {}}
          focusDelay={0}
          getDocument={[Function]}
          getPopupClassNameFromAlign={[Function]}
          hideAction={
            Array [
              "click",
            ]
          }
          inputValue=""
          mask={false}
          maskClosable={true}
          mouseEnterDelay={0}
          mouseLeaveDelay={0.1}
          onDropdownVisibleChange={[Function]}
          onMenuDeselect={[Function]}
          onMenuSelect={[Function]}
          onPopupAlign={[Function]}
          onPopupVisibleChange={[Function]}
          options={Array []}
          popup={
            <DropdownMenu
              defaultActiveFirstOption={true}
              dropdownMenuStyle={Object {}}
              firstActiveValue={undefined}
              inputValue=""
              menuItems={Array []}
              multiple={undefined}
              onMenuDeselect={[Function]}
              onMenuSelect={[Function]}
              onPopupFocus={[Function]}
              prefixCls="ant-select-dropdown"
              value={Array []}
              visible={false}
            />
          }
          popupAlign={Object {}}
          popupClassName="ant-select-dropdown--single"
          popupPlacement="bottomLeft"
          popupStyle={Object {}}
          popupTransitionName="slide-up"
          popupVisible={false}
          prefixCls="ant-select-dropdown"
          showAction={
            Array [
              "click",
            ]
          }
          showSearch={false}
          transitionName="slide-up"
          value={Array []}
          visible={false}
        >
          <div
            className="ant-select-lg ant-select ant-select-enabled"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseDown={[Function]}
            onTouchStart={[Function]}
          >
            <div
              aria-autocomplete="list"
              aria-expanded={false}
              aria-haspopup="true"
              className="ant-select-selection
            ant-select-selection--single"
              key="selection"
              onKeyDown={[Function]}
              role="combobox"
              tabIndex={0}
            >
              <div
                className="ant-select-selection__rendered"
              />
              <span
                className="ant-select-arrow"
                key="arrow"
                onClick={[Function]}
                style={
                  Object {
                    "WebkitUserSelect": "none",
                    "userSelect": "none",
                  }
                }
                unselectable="unselectable"
              >
                <b />
              </span>
            </div>
          </div>
        </Trigger>
      </SelectTrigger>
    </Select>
  </Select>,
  <Select
    allowClear={false}
    backfill={false}
    choiceTransitionName="zoom"
    className="ant-select-lg"
    defaultActiveFirstOption={true}
    defaultOpen={false}
    dropdownMatchSelectWidth={true}
    dropdownMenuStyle={Object {}}
    dropdownStyle={Object {}}
    labelInValue={false}
    notFoundContent="无匹配结果"
    onBlur={[Function]}
    onChange={[Function]}
    onDeselect={[Function]}
    onFocus={[Function]}
    onSearch={[Function]}
    onSelect={[Function]}
    optionFilterProp="value"
    optionLabelProp="children"
    placeholder=""
    prefixCls="ant-select"
    showArrow={true}
    showSearch={false}
    transitionName="slide-up"
  >
    <SelectTrigger
      defaultActiveFirstOption={true}
      dropdownMatchSelectWidth={true}
      dropdownMenuStyle={Object {}}
      dropdownStyle={Object {}}
      inputValue=""
      onDropdownVisibleChange={[Function]}
      onMenuDeselect={[Function]}
      onMenuSelect={[Function]}
      onPopupFocus={[Function]}
      options={Array []}
      prefixCls="ant-select"
      showSearch={false}
      transitionName="slide-up"
      value={Array []}
      visible={false}
    >
      <Trigger
        action={Array []}
        afterPopupVisibleChange={[Function]}
        blurDelay={0.15}
        builtinPlacements={
          Object {
            "bottomLeft": Object {
              "offset": Array [
                0,
                4,
              ],
              "overflow": Object {
                "adjustX": 0,
                "adjustY": 1,
              },
              "points": Array [
                "tl",
                "bl",
              ],
            },
            "topLeft": Object {
              "offset": Array [
                0,
                -4,
              ],
              "overflow": Object {
                "adjustX": 0,
                "adjustY": 1,
              },
              "points": Array [
                "bl",
                "tl",
              ],
            },
          }
        }
        defaultActiveFirstOption={true}
        defaultPopupVisible={false}
        destroyPopupOnHide={false}
        dropdownMatchSelectWidth={true}
        dropdownMenuStyle={Object {}}
        dropdownStyle={Object {}}
        focusDelay={0}
        getDocument={[Function]}
        getPopupClassNameFromAlign={[Function]}
        hideAction={
          Array [
            "click",
          ]
        }
        inputValue=""
        mask={false}
        maskClosable={true}
        mouseEnterDelay={0}
        mouseLeaveDelay={0.1}
        onDropdownVisibleChange={[Function]}
        onMenuDeselect={[Function]}
        onMenuSelect={[Function]}
        onPopupAlign={[Function]}
        onPopupVisibleChange={[Function]}
        options={Array []}
        popup={
          <DropdownMenu
            defaultActiveFirstOption={true}
            dropdownMenuStyle={Object {}}
            firstActiveValue={undefined}
            inputValue=""
            menuItems={Array []}
            multiple={undefined}
            onMenuDeselect={[Function]}
            onMenuSelect={[Function]}
            onPopupFocus={[Function]}
            prefixCls="ant-select-dropdown"
            value={Array []}
            visible={false}
          />
        }
        popupAlign={Object {}}
        popupClassName="ant-select-dropdown--single"
        popupPlacement="bottomLeft"
        popupStyle={Object {}}
        popupTransitionName="slide-up"
        popupVisible={false}
        prefixCls="ant-select-dropdown"
        showAction={
          Array [
            "click",
          ]
        }
        showSearch={false}
        transitionName="slide-up"
        value={Array []}
        visible={false}
      >
        <div
          className="ant-select-lg ant-select ant-select-enabled"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
        >
          <div
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup="true"
            className="ant-select-selection
            ant-select-selection--single"
            key="selection"
            onKeyDown={[Function]}
            role="combobox"
            tabIndex={0}
          >
            <div
              className="ant-select-selection__rendered"
            />
            <span
              className="ant-select-arrow"
              key="arrow"
              onClick={[Function]}
              style={
                Object {
                  "WebkitUserSelect": "none",
                  "userSelect": "none",
                }
              }
              unselectable="unselectable"
            >
              <b />
            </span>
          </div>
        </div>
      </Trigger>
    </SelectTrigger>
  </Select>,
]
`;
